crates/ruff_linter/src/rules/flake8_pyi/rules/bytestring_usage.rs

```diff
@@ -0,0 +1,83 @@
+use ruff_diagnostics::{Diagnostic, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, Expr};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for the uses of `typing.ByteString` and `collections.abc.ByteString`.
+///
+/// ## Why is this bad?
+/// `ByteString` has been deprecated since Python 3.9 and will be removed in
+/// Python 3.14. The Python documentation recommends using either
+/// `collections.abc.Buffer` (for Python >=3.12) or a union like
+/// `bytes | bytearray | memoryview` instead.
+///
+/// ## Example
+/// ```python
+/// from typing import ByteString
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from collections.abc import Buffer
+/// ```
+///
+/// ## References
+/// - [Python documentation: The `ByteString` type](https://docs.python.org/3/library/typing.html#typing.ByteString)
+#[violation]
+pub struct ByteStringUsage {
+    full_name: String,
+}
+
+impl Violation for ByteStringUsage {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::None;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let ByteStringUsage { full_name } = self;
+        format!("Do not use `{full_name}`, which has unclear semantics and is deprecated")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let ByteStringUsage { full_name } = self;
+        Some(format!("Do not use `{full_name}`"))
+    }
```

Comment:
This message should be a succinct summary of what the user should do to fix the violation (if an autofix is available, and you're using e.g. VSCode, double-clicking on this message when you hover over the violation in the editor will apply the autofix).

In this case I think I'd probably not implement this method at all! Since we're not offering an autofix, it's not necessary. And it's hard for us to tell exactly what the best fix is in any one case -- in some cases it might be to use `collections.abc.Buffer`, in some cases it might be to use `typing_extensions.Buffer`, and in some cases it might be to use a union. (The ambiguity is precisely why we've deprecated `ByteString`!)

```suggestion
```